lib/erlef/session.ex

```diff
@@ -0,0 +1,201 @@
+defmodule Erlef.Session do
+  @moduledoc """
+  Erlef.Session - Session handler and session serializer/deserializer
+  """
+
+  @enforce_keys [:account_id, :access_token, :refresh_token, :username, :expires_at, :is_admin]
+  defstruct [:account_id, :access_token, :refresh_token, :username, :expires_at, :is_admin]
+
+  @type auth_data() :: map()
+
+  @type t() :: %__MODULE__{
+          account_id: integer(),
+          access_token: String.t(),
+          refresh_token: String.t(),
+          username: String.t(),
+          expires_at: String.t(),
+          is_admin: boolean()
+        }
+
+  @data_map %{
+    "account_id" => :account_id,
+    "access_token" => :access_token,
+    "refresh_token" => :refresh_token,
+    "username" => :username,
+    "expires_at" => :expires_at,
+    "is_admin" => :is_admin
+  }
+
+  @spec new(String.t(), map()) :: t()
+  def new(username, %{"expires_in" => expires_in} = data) do
+    %{
+      "Permissions" => [%{"AccountId" => aid} | _],
+      "access_token" => access_token,
+      "refresh_token" => refresh_token
+    } = data
+
+    case is_admin(aid, access_token) do
+      truth when is_boolean(truth) ->
+        %__MODULE__{
+          account_id: aid,
+          access_token: access_token,
+          refresh_token: refresh_token,
+          username: username,
+          expires_at: expires_at(expires_in),
+          is_admin: truth
+        }
+
+      err ->
+        err
+    end
+  end
+
+  @spec normalize(map()) :: t() | nil
+  def normalize(nil), do: nil
+
+  def normalize(data) do
+    normal =
+      Enum.reduce(data, %{}, fn {k, v}, acc ->
+        case @data_map[k] do
+          nil ->
+            acc
+
+          key ->
+            Map.put(acc, key, v)
+        end
+      end)
+
+    struct(__MODULE__, normal)
```

Comment:
If this is the re-serialization, you will want to explicitly unset the `is_admin` value and re-check it every time, unless you're able to sign the payload. Otherwise anyone can just set the `is_admin` flag by hand and grant themselves access. 

If `struct/2` does forward to `new/2` you can disregard this (I don't recall the exact rules) but the implementation of `new()` does not make it sound like it'd be doing that automatically since the type of struct to cast has to be passed...